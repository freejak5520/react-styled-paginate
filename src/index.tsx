import React from 'react';

const DEFAULT_PAGE_SIZE = 5;

type PageButtonComponentType = React.FC<
  React.HTMLAttributes<HTMLDivElement> & {
    active?: boolean;
    disabled?: boolean;
  }
>;

const Container = ({ children }: { children: React.ReactNode }) => {
  return <div style={{ display: 'flex', gap: 6 }}>{children}</div>;
};

const PageButtonComponent: PageButtonComponentType = ({
  children,
  onClick,
  active,
  disabled,
  ...rest
}) => {
  return (
    <div
      onClick={e => {
        if (disabled || active) return;
        onClick && onClick(e);
      }}
      {...rest}
      style={{
        padding: '4px 8px',
        background: active ? '#123456' : '',
        color: active ? 'white' : '',
        cursor: disabled ? 'not-allowed' : 'pointer',
      }}
    >
      {children}
    </div>
  );
};

const components = {
  pageButtonComponent: PageButtonComponent,
  container: Container,
};

type PaginateProps = {
  page?: number;
  totalPages: number;
  pageSize?: number;
  onChange?: (page: number) => void;
  containerComponent?: React.FC<{ children: React.ReactNode }>;
  pageButtonComponent?: PageButtonComponentType;
};

export const Paginate = ({
  page = 1,
  totalPages,
  pageSize = DEFAULT_PAGE_SIZE,
  onChange,
  containerComponent = components.container,
  pageButtonComponent = components.pageButtonComponent,
}: PaginateProps) => {
  const Container = containerComponent;
  const PageButtonComponent = pageButtonComponent;

  const calculatePageOffsets = (
    page: number,
    pageSize: number,
    totalPages: number
  ) => {
    let startOffset = page - Math.floor(pageSize / 2 - 0.1);
    let endOffset = page + Math.floor(pageSize / 2);

    let startPage = Math.max(1, startOffset);
    if (totalPages < endOffset) {
      const diff = endOffset - totalPages;
      startPage -= diff;
    }

    const endPage = Math.min(
      totalPages,
      endOffset + (startOffset <= 0 ? -startOffset + 1 : 0)
    );
    return { startPage, endPage };
  };

  const makePageNumberArray = (startPage: number, endPage: number) => {
    return Array.from(
      { length: endPage - startPage + 1 },
      (_, index) => startPage + index
    );
  };

  const { startPage, endPage } = calculatePageOffsets(
    page,
    pageSize,
    totalPages
  );

  return (
    <Container>
      <PageButtonComponent disabled={page === 1} onClick={() => onChange?.(1)}>
        First
      </PageButtonComponent>
      <PageButtonComponent
        disabled={page === 1}
        onClick={() => onChange?.(page - 1)}
      >
        Previous
      </PageButtonComponent>
      {makePageNumberArray(startPage, endPage).map(pageNumber => (
        <PageButtonComponent
          key={pageNumber}
          onClick={() => onChange?.(pageNumber)}
          active={page === pageNumber}
        >
          {pageNumber}
        </PageButtonComponent>
      ))}
      <PageButtonComponent
        disabled={page === totalPages}
        onClick={() => onChange?.(page + 1)}
      >
        Next
      </PageButtonComponent>
      <PageButtonComponent
        disabled={page === totalPages}
        onClick={() => onChange?.(totalPages)}
      >
        Last
      </PageButtonComponent>
    </Container>
  );
};
