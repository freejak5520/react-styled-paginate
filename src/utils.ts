export const calculatePageOffsets = (
  currentPage: number,
  perPage: number,
  total: number
) => {
  let startOffset = currentPage - Math.floor(perPage / 2 - 0.1);
  let endOffset = currentPage + Math.floor(perPage / 2);

  let startPage = Math.max(1, startOffset);
  if (total < endOffset) {
    const diff = endOffset - total;
    startPage -= diff;
  }

  const endPage = Math.min(
    total,
    endOffset + (startOffset <= 0 ? -startOffset + 1 : 0)
  );
  return { startPage, endPage };
};
