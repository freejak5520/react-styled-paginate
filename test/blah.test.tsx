import React from 'react';
import renderer from 'react-test-renderer';
import { Paginate } from '../src';

describe('it', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Paginate totalPages={100} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with custom component', () => {
    const component = renderer.create(
      <Paginate totalPages={100} pageSize={20} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
