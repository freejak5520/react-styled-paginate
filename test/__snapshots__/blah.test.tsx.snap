// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders with custom component 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "gap": 6,
    }
  }
>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "not-allowed",
        "padding": "4px 8px",
      }
    }
  >
    First
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "not-allowed",
        "padding": "4px 8px",
      }
    }
  >
    Previous
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "#123456",
        "color": "white",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    1
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    2
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    3
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    4
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    5
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    6
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    7
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    8
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    9
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    10
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    11
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    12
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    13
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    14
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    15
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    16
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    17
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    18
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    19
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    20
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    Next
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    Last
  </div>
</div>
`;

exports[`it renders without crashing 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "gap": 6,
    }
  }
>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "not-allowed",
        "padding": "4px 8px",
      }
    }
  >
    First
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "not-allowed",
        "padding": "4px 8px",
      }
    }
  >
    Previous
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "#123456",
        "color": "white",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    1
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    2
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    3
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    4
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    5
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    Next
  </div>
  <div
    onClick={[Function]}
    style={
      Object {
        "background": "",
        "color": "",
        "cursor": "pointer",
        "padding": "4px 8px",
      }
    }
  >
    Last
  </div>
</div>
`;
